File: src/pieces/KingBehaviour.rs
use crate::{
    array::get_surrounding_positions,
    types::{Array2D, Position, PositionContent, Shape},
};

use super::piece::Piece;

pub(crate) fn allowed_moves(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let surrounding_positions = get_surrounding_positions(pos, shape);
    let mut filt_all_ms: Vec<Position> = Vec::with_capacity(surrounding_positions.len());
    for move_pos in surrounding_positions.into_iter() {
        let mp_piece: &PositionContent =
            &board[usize::try_from(move_pos.0).unwrap()][usize::try_from(move_pos.1).unwrap()];
        match mp_piece {
            PositionContent::PieceContent(_) => (),
            PositionContent::Empty => filt_all_ms.push(move_pos),
        }
    }
    filt_all_ms
}

pub(crate) fn allowed_takes(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let surrounding_positions = get_surrounding_positions(pos, shape);
    let mut filt_all_ts: Vec<Position> = Vec::with_capacity(surrounding_positions.len());
    for take_pos in surrounding_positions.into_iter() {
        let tp_piece: &PositionContent =
            &board[usize::try_from(take_pos.0).unwrap()][usize::try_from(take_pos.1).unwrap()];
        match tp_piece {
            PositionContent::PieceContent(_) => filt_all_ts.push(take_pos),
            PositionContent::Empty => (),
        }
    }
    filt_all_ts
}

File: src/pieces/KingBehaviour.ts
import { Array2D, get_surrounding_positions } from "../Array"
import { Piece } from "./Piece"

export function king_allowed_moves(piece: Piece, board: Array2D, pos: Position, is_white: boolean): Position[] {
    const surrounding_positions = get_surrounding_positions(pos, board.shape)
    const filt_all_ms: Position[] = []
    for (const move_pos of surrounding_positions) {
        const mp_piece = board.data[move_pos[0]][move_pos[1]]
        if (mp_piece.is_empty) {
            filt_all_ms.push(move_pos)
        }
    }
    return filt_all_ms
}

export function king_allowed_takes(piece: Piece, board: Array2D, pos: Position, is_white: boolean): Position[] {
    const surrounding_positions = get_surrounding_positions(pos, board.shape)
    const filt_all_ts: Position[] = []
    for (const take_pos of surrounding_positions) {
        const tp_piece = board.data[take_pos[0]][take_pos[1]]
        if (tp_piece.is_white != piece.is_white){
            filt_all_ts.push(take_pos)
        }
    }
    return filt_all_ts
}